File: server/app/domain/export/csv_engine.py
"""
A generic, DTO-agnostic CSV engine (Phase 7E, Section 8/9) - built on Python's standard `csv`
module (RFC 4180 quoting/escaping is handled by the stdlib, never hand-rolled string
concatenation). Every monetary value passed in must already be a paise integer (as a string) -
this module never converts anything to a float; that discipline is enforced by the caller
(`export_service.py`'s dict shapes only ever carry `...Paise` integer fields).
"""
from __future__ import annotations

import csv
import io


def write_csv(headers: list[str], rows: list[list]) -> str:
    """[rows] must all have the same length as [headers]. Row/column order is exactly the
    caller's own list order - never re-sorted. Uses `\r\n` line endings per RFC 4180."""
    buffer = io.StringIO()
    writer = csv.writer(buffer, lineterminator="\r\n")
    writer.writerow(headers)
    for row in rows:
        writer.writerow(["" if v is None else v for v in row])
    return buffer.getvalue()


def voucher_csv(voucher: dict) -> str:
    headers = [
        "voucherId", "voucherNumber", "voucherType", "date", "referenceNumber", "narration",
        "totalAmountPaise", "isPosted", "isCancelled", "ledgerId", "ledgerName", "drCr", "lineAmountPaise", "lineNarration", "lineOrder",
    ]
    rows = [
        [
            voucher["voucherId"], voucher["voucherNumber"], voucher["voucherType"], voucher["date"],
            voucher["referenceNumber"], voucher["narration"], voucher["totalAmountPaise"], voucher["isPosted"], voucher["isCancelled"],
            line["ledgerId"], line["ledgerName"], line["type"], line["amountPaise"], line["narration"], line["lineOrder"],
        ]
        for line in voucher["journalLines"]
    ]
    return write_csv(headers, rows)


def parties_csv(parties: list[dict]) -> str:
    headers = ["partyId", "ledgerId", "role", "entityType", "displayName", "contactName", "creditLimitPaise", "paymentTerms", "isActive"]
    rows = [[p["partyId"], p["ledgerId"], p["role"], p["entityType"], p["displayName"], p["contactName"], p["creditLimitPaise"], p["paymentTerms"], p["isActive"]] for p in parties]
    return write_csv(headers, rows)


def ledgers_csv(ledgers: list[dict]) -> str:
    headers = ["ledgerId", "groupId", "name", "code", "openingBalancePaise", "openingBalanceType", "currentBalancePaise", "currentBalanceType", "gstin", "pan", "stateCode", "address", "isSystem", "isActive"]
    rows = [
        [l["ledgerId"], l["groupId"], l["name"], l["code"], l["openingBalancePaise"], l["openingBalanceType"], l["currentBalancePaise"], l["currentBalanceType"], l["gstin"], l["pan"], l["stateCode"], l["address"], l["isSystem"], l["isActive"]]
        for l in ledgers
    ]
    return write_csv(headers, rows)


def trial_balance_csv(dto: dict) -> str:
    headers = ["ledgerId", "ledgerName", "groupId", "groupName", "primaryGroup", "openingDebitPaise", "openingCreditPaise", "transactionDebitPaise", "transactionCreditPaise", "closingDebitPaise", "closingCreditPaise"]
    rows = [
        [r["ledgerId"], r["ledgerName"], r["groupId"], r["groupName"], r["primaryGroup"], r["openingDebitPaise"], r["openingCreditPaise"], r["transactionDebitPaise"], r["transactionCreditPaise"], r["closingDebitPaise"], r["closingCreditPaise"]]
        for r in dto["rows"]
    ]
    return write_csv(headers, rows)


def outstanding_csv(dto: dict) -> str:
    headers = ["invoiceId", "invoiceNumber", "invoiceType", "partyId", "partyName", "voucherNumber", "date", "dueDate", "totalAmountPaise", "outstandingAmountPaise", "status", "daysOutstanding", "agingBucket"]
    rows = [
        [r["invoiceId"], r["invoiceNumber"], r["invoiceType"], r["partyId"], r["partyName"], r["voucherNumber"], r["date"], r["dueDate"], r["totalAmountPaise"], r["outstandingAmountPaise"], r["status"], r["daysOutstanding"], r["agingBucket"]]
        for r in dto["rows"]
    ]
    return write_csv(headers, rows)


def gst_summary_csv(dto: dict) -> str:
    headers = ["companyName", "gstin", "period", "totalTaxableOutwardPaise", "totalTaxOutwardPaise", "totalTaxableInwardPaise", "totalTaxInwardItcPaise", "netTaxPayablePaise", "totalCessPaise", "netCessPayablePaise"]
    row = [dto["companyName"], dto["gstin"], dto["period"], dto["totalTaxableOutwardPaise"], dto["totalTaxOutwardPaise"], dto["totalTaxableInwardPaise"], dto["totalTaxInwardItcPaise"], dto["netTaxPayablePaise"], dto["totalCessPaise"], dto["netCessPayablePaise"]]
    return write_csv(headers, [row])


def gst_transactions_csv(transactions: list[dict]) -> str:
    headers = ["gstTransactionId", "voucherId", "voucherType", "partyGstin", "placeOfSupply", "supplyType", "hsnSacCode", "isService", "taxableAmountPaise", "gstRatePercent", "cgstPaise", "sgstPaise", "igstPaise", "cessPaise", "direction", "lineOrder"]
    rows = [
        [t["gstTransactionId"], t["voucherId"], t["voucherType"], t["partyGstin"], t["placeOfSupply"], t["supplyType"], t["hsnSacCode"], t["isService"], t["taxableAmountPaise"], t["gstRatePercent"], t["cgstPaise"], t["sgstPaise"], t["igstPaise"], t["cessPaise"], t["direction"], t["lineOrder"]]
        for t in transactions
    ]
    return write_csv(headers, rows)

File: app/src/main/java/com/example/accounting/domain/export/CsvEngine.kt
package com.example.accounting.domain.export

/**
 * A generic, report/DTO-agnostic CSV engine (Phase 7E, Section 8/9) - distinct from and never
 * modifying `domain.rendering.CsvExporter` (Phase 7D, frozen, document-line-specific). Every
 * monetary value is written as an exact paise `Long` (never `.toRupeesDouble()`/floating-point,
 * unlike the 7D exporter it deliberately does not reuse) - Section 7's "never export a
 * floating-point representation as the authoritative amount" applies to CSV exactly as it does to
 * JSON. Handles commas/quotes/newlines/Unicode/empty/null per RFC 4180; column and row order are
 * always deterministic (the caller's own list order - never re-sorted).
 */
object CsvEngine {
    /** [rows] must all have the same length as [headers] - callers build rows via each DTO's own
     * `toCsvRow()`/`toCsvRows()` extension, never hand-rolled string concatenation. */
    fun write(headers: List<String>, rows: List<List<String?>>): String {
        val sb = StringBuilder()
        sb.append(headers.joinToString(",") { field(it) }).append("\r\n")
        for (row in rows) {
            sb.append(row.joinToString(",") { field(it) }).append("\r\n")
        }
        return sb.toString()
    }

    /** RFC 4180 field escaping: null/empty become an empty field; a field containing a comma,
     * double-quote, or CR/LF is wrapped in quotes with internal quotes doubled. Unicode passes
     * through untouched (no transliteration/encoding narrowing). */
    private fun field(value: String?): String {
        if (value.isNullOrEmpty()) return ""
        return if (value.any { it == ',' || it == '"' || it == '\n' || it == '\r' }) {
            "\"${value.replace("\"", "\"\"")}\""
        } else value
    }
}

// ==================== Per-DTO CSV row mappers (deterministic column order) ====================

fun VoucherExportDto.toCsvHeaders(): List<String> = listOf(
    "voucherId", "voucherNumber", "voucherType", "date", "referenceNumber", "narration",
    "totalAmountPaise", "isPosted", "isCancelled", "ledgerId", "ledgerName", "drCr", "lineAmountPaise", "lineNarration", "lineOrder"
)

/** One row per journal line (a voucher with N lines produces N CSV rows, all sharing the voucher's
 * own header columns) - this is the standard "denormalized ledger export" shape accounting tools
 * expect, not a single summarized row that would lose per-line detail. */
fun VoucherExportDto.toCsvRows(): List<List<String?>> = journalLines.map { line ->
    listOf(
        voucherId, voucherNumber, voucherType.name, date.toString(), referenceNumber, narration,
        totalAmountPaise.toString(), isPosted.toString(), isCancelled.toString(),
        line.ledgerId, line.ledgerName, line.type.name, line.amountPaise.toString(), line.narration, line.lineOrder.toString()
    )
}

// Named distinctly (not `List<T>.toCsvHeaders()`/`toCsvRows()`) - generic extension functions on
// `List<T>` erase to the same JVM signature regardless of `T` ("platform declaration clash"),
// so these can't be overloaded the way the per-DTO-type extensions above are.
fun List<PartyExportDto>.toPartyCsvHeaders(): List<String> = listOf(
    "partyId", "ledgerId", "role", "entityType", "displayName", "contactName", "creditLimitPaise", "paymentTerms", "isActive"
)

fun List<PartyExportDto>.toPartyCsvRows(): List<List<String?>> = map {
    listOf(it.partyId, it.ledgerId, it.role, it.entityType, it.displayName, it.contactName, it.creditLimitPaise?.toString(), it.paymentTerms, it.isActive.toString())
}

fun List<LedgerExportDto>.toLedgerCsvHeaders(): List<String> = listOf(
    "ledgerId", "groupId", "name", "code", "openingBalancePaise", "openingBalanceType",
    "currentBalancePaise", "currentBalanceType", "gstin", "pan", "stateCode", "address", "isSystem", "isActive"
)

fun List<LedgerExportDto>.toLedgerCsvRows(): List<List<String?>> = map {
    listOf(
        it.ledgerId, it.groupId, it.name, it.code, it.openingBalancePaise.toString(), it.openingBalanceType.name,
        it.currentBalancePaise.toString(), it.currentBalanceType.name, it.gstin, it.pan, it.stateCode, it.address,
        it.isSystem.toString(), it.isActive.toString()
    )
}

fun TrialBalanceExportDto.toCsvHeaders(): List<String> = listOf(
    "ledgerId", "ledgerName", "groupId", "groupName", "primaryGroup",
    "openingDebitPaise", "openingCreditPaise", "transactionDebitPaise", "transactionCreditPaise", "closingDebitPaise", "closingCreditPaise"
)

fun TrialBalanceExportDto.toCsvRows(): List<List<String?>> = rows.map {
    listOf(
        it.ledgerId, it.ledgerName, it.groupId, it.groupName, it.primaryGroup.name,
        it.openingDebitPaise.toString(), it.openingCreditPaise.toString(), it.transactionDebitPaise.toString(),
        it.transactionCreditPaise.toString(), it.closingDebitPaise.toString(), it.closingCreditPaise.toString()
    )
}

fun ProfitAndLossExportDto.toCsvHeaders(): List<String> = listOf(
    "companyName", "financialYearCode", "dateRange", "salesRevenuePaise", "directIncomesPaise",
    "purchasesPaise", "directExpensesPaise", "grossProfitPaise", "indirectIncomesPaise", "indirectExpensesPaise",
    "netProfitPaise", "isInventoryAware"
)

fun ProfitAndLossExportDto.toCsvRows(): List<List<String?>> = listOf(
    listOf(
        companyName, financialYearCode, dateRange, salesRevenuePaise.toString(), directIncomesPaise.toString(),
        purchasesPaise.toString(), directExpensesPaise.toString(), grossProfitPaise.toString(),
        indirectIncomesPaise.toString(), indirectExpensesPaise.toString(), netProfitPaise.toString(), isInventoryAware.toString()
    )
)

fun BalanceSheetExportDto.toCsvHeaders(): List<String> = listOf(
    "companyName", "financialYearCode", "asOfDate", "totalLiabilitiesPaise", "totalAssetsPaise", "isBalanced",
    "capitalAccountsPaise", "loansLiabilitiesPaise", "currentLiabilitiesPaise", "fixedAssetsPaise",
    "currentAssetsPaise", "sundryDebtorsPaise", "bankAccountsPaise", "cashInHandPaise", "stockInHandPaise"
)

fun BalanceSheetExportDto.toCsvRows(): List<List<String?>> = listOf(
    listOf(
        companyName, financialYearCode, asOfDate.toString(), totalLiabilitiesPaise.toString(), totalAssetsPaise.toString(), isBalanced.toString(),
        capitalAccountsPaise.toString(), loansLiabilitiesPaise.toString(), currentLiabilitiesPaise.toString(), fixedAssetsPaise.toString(),
        currentAssetsPaise.toString(), sundryDebtorsPaise.toString(), bankAccountsPaise.toString(), cashInHandPaise.toString(), stockInHandPaise.toString()
    )
)

fun OutstandingExportDto.toCsvHeaders(): List<String> = listOf(
    "invoiceId", "invoiceNumber", "invoiceType", "partyId", "partyName", "voucherNumber",
    "date", "dueDate", "totalAmountPaise", "outstandingAmountPaise", "status", "daysOutstanding", "agingBucket"
)

fun OutstandingExportDto.toCsvRows(): List<List<String?>> = rows.map {
    listOf(
        it.invoiceId, it.invoiceNumber, it.invoiceType.name, it.partyId, it.partyName, it.voucherNumber,
        it.date.toString(), it.dueDate?.toString(), it.totalAmountPaise.toString(), it.outstandingAmountPaise.toString(),
        it.status.name, it.daysOutstanding.toString(), it.agingBucket.name
    )
}

fun GSTSummaryExportDto.toCsvHeaders(): List<String> = listOf(
    "companyName", "gstin", "period", "totalTaxableOutwardPaise", "totalTaxOutwardPaise",
    "totalTaxableInwardPaise", "totalTaxInwardItcPaise", "netTaxPayablePaise", "totalCessPaise", "netCessPayablePaise"
)

fun GSTSummaryExportDto.toCsvRows(): List<List<String?>> = listOf(
    listOf(
        companyName, gstin, period, totalTaxableOutwardPaise.toString(), totalTaxOutwardPaise.toString(),
        totalTaxableInwardPaise.toString(), totalTaxInwardItcPaise.toString(), netTaxPayablePaise.toString(),
        totalCessPaise.toString(), netCessPayablePaise.toString()
    )
)

fun List<GSTTransactionExportDto>.toGstTransactionCsvHeaders(): List<String> = listOf(
    "gstTransactionId", "voucherId", "voucherType", "partyGstin", "placeOfSupply", "supplyType",
    "hsnSacCode", "isService", "taxableAmountPaise", "gstRatePercent", "cgstPaise", "sgstPaise", "igstPaise", "cessPaise", "direction", "lineOrder"
)

fun List<GSTTransactionExportDto>.toGstTransactionCsvRows(): List<List<String?>> = map {
    listOf(
        it.gstTransactionId, it.voucherId, it.voucherType.name, it.partyGstin, it.placeOfSupply, it.supplyType,
        it.hsnSacCode, it.isService?.toString(), it.taxableAmountPaise.toString(), it.gstRatePercent.toString(),
        it.cgstPaise.toString(), it.sgstPaise.toString(), it.igstPaise.toString(), it.cessPaise.toString(),
        it.direction, it.lineOrder.toString()
    )
}
